### PR TITLE
GCE Firewall (Ready for Review)

### PR DESCRIPTION
**Pull request by ArtfulCoder**
# K8S-GCE Firewall

The K8S-GCE Firewall has two components, client and server. 
## Server

The server serves GCE Firewall configuration to clients. The server itself gets the configuration from Google Cloud. The GCE/GKE cluster needs to be created with the right permissions set so that instances can read the cluster configuration using Google Cloud API. 
The server polls for firewall configuration changes. Each time the server has new firewall configuration, it pushes the new configuration to clients.
## Clients

K8S-GCE Client runs on every node of the cluster, except the master. The client discovers the server using the server's in-cluster DNS name and maintains a persistent websocket connection with the server.
When the client receives a new config, it creates corresponding iptable changes to reflect configuration.
## Configuration
- A user can configure the GCE Firewall  via the description field of a firewall rule.
  The description field should look like:
  KubernetesAutoGenerated_OnlyAllowTrafficForDestinationIP_104.197.89.161
  This means that only traffic for destination ip 104.197.89.161 will be allowed on ports opened by that firewall rule.
- The K8S-GCE firewall always allows packets with local source ips.
- The K8S-GCE firewall always allows packets with source has local ips.
- The K8S-GCE firewall always allows packets on established connections. 
- The K8S-GCE firewall mirrors firewall rules created by the user on GCE.
- The default K8S-GCE firewall rule is to drop all packets.
### GCE Firewall Rules

firewall
**Click on image to get enlarged view**

kubetest-firewall-942320723 has the following description: 
KubernetesAutoGenerated_OnlyAllowTrafficForDestinationIP_104.197.89.161
### Corresponding iptables rules

*mangle
:K8S-FIREWALL - [0:0]
-A PREROUTING -j K8S-FIREWALL
-A K8S-FIREWALL -d 104.197.89.1612/32 -p tcp -m comment --comment kubetest-firewall-942320723 -m tcp --dport 8000 -j MARK --set-xmark 0x70617373/0xffffffff
COMMIT
*filter
:INPUT ACCEPT [291:88151]
:FORWARD ACCEPT [124:175178]
:OUTPUT ACCEPT [248:567014]
:DOCKER - [0:0]
:K8S-FIREWALL - [0:0]
-A INPUT -j K8S-FIREWALL
-A FORWARD -o docker0 -j DOCKER
-A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i docker0 ! -o docker0 -j ACCEPT
-A FORWARD -i docker0 -o docker0 -j ACCEPT
-A K8S-FIREWALL -m comment --comment allow-outbound-traffic -m conntrack --ctstate RELATED,ESTABLISHED -j RETURN
-A K8S-FIREWALL -m comment --comment allow-local-src-addreses -m addrtype --src-type LOCAL -j RETURN
-A K8S-FIREWALL -s 10.0.0.0/8 -m comment --comment "allow-10.0.0.0/8-addresses" -j RETURN
-A K8S-FIREWALL -s 169.254.0.0/16 -m comment --comment "allow-169.254.0.0/16-addresses" -j RETURN
-A K8S-FIREWALL -s 172.16.0.0/12 -m comment --comment "allow-172.16.0.0/12-addresses" -j RETURN
-A K8S-FIREWALL -s 192.168.0.0/16 -m comment --comment "allow-192.168.0.0/16-addresses" -j RETURN
-A K8S-FIREWALL -s 10.0.0.0/8 -p tcp -m comment --comment default-default-internal -m tcp --dport 1:65535 -j RETURN
-A K8S-FIREWALL -s 10.0.0.0/8 -p udp -m comment --comment default-default-internal -m udp --dport 1:65535 -j RETURN
-A K8S-FIREWALL -s 10.0.0.0/8 -p icmp -m comment --comment default-default-internal -j RETURN
-A K8S-FIREWALL -p tcp -m comment --comment default-default-ssh -m tcp --dport 22 -j RETURN
-A K8S-FIREWALL -s 10.0.0.0/8 -p tcp -m comment --comment e2e-default-internal -m tcp --dport 1:65535 -j RETURN
-A K8S-FIREWALL -s 10.0.0.0/8 -p udp -m comment --comment e2e-default-internal -m udp --dport 1:65535 -j RETURN
-A K8S-FIREWALL -s 10.0.0.0/8 -p icmp -m comment --comment e2e-default-internal -j RETURN
-A K8S-FIREWALL -p tcp -m comment --comment e2e-default-ssh -m tcp --dport 22 -j RETURN
-A K8S-FIREWALL -s 10.245.0.0/16 -p tcp -m comment --comment e2e-test-abshah-minion-all -j RETURN
-A K8S-FIREWALL -s 10.245.0.0/16 -p udp -m comment --comment e2e-test-abshah-minion-all -j RETURN
-A K8S-FIREWALL -s 10.245.0.0/16 -p icmp -m comment --comment e2e-test-abshah-minion-all -j RETURN
-A K8S-FIREWALL -s 10.245.0.0/16 -p esp -m comment --comment e2e-test-abshah-minion-all -j RETURN
-A K8S-FIREWALL -s 10.245.0.0/16 -p ah -m comment --comment e2e-test-abshah-minion-all -j RETURN
-A K8S-FIREWALL -s 10.245.0.0/16 -p sctp -m comment --comment e2e-test-abshah-minion-all -j RETURN
-A K8S-FIREWALL -p tcp -m comment --comment e2e-test-abshah-minion-e2e-test-abshah-nodeports -m tcp --dport 30000:32767 -j RETURN
-A K8S-FIREWALL -p udp -m comment --comment e2e-test-abshah-minion-e2e-test-abshah-nodeports -m udp --dport 30000:32767 -j RETURN
-A K8S-FIREWALL -p tcp -m comment --comment pprof -m tcp --dport 10249 -j RETURN
-A K8S-FIREWALL -m comment --comment k8s-forwarding-rule-traffic -m mark --mark 0x70617373 -j RETURN
-A K8S-FIREWALL -j DROP
COMMIT

### CODE REVIEW

**Comment by ArtfulCoder** [comment]
manually verified that the code works.
e2e test needs to be added.
Also blocked on #17406

**Comment by eparis** [comment]
And here I was just about to un-assign myself.  Thank you @ArtfulCoder

**Comment by smarterclayton** [comment]
Kube aware firewall apply daemon?  Takes GCE firewall rules and maps
them to each node in the cluster?

**Comment by ArtfulCoder** [comment]
@smarterclayton  that is correct.
The PR is ready for review now.

**Comment by ArtfulCoder** [comment]
The docker image hosting would need to be in gcr.io, once the PR is LGTM-ed

File: pkg/util/iptables/iptables.go [hunk 1]
```diff
@@ -283,6 +304,18 @@ func (runner *runner) DeleteRule(table Table, chain Chain, args ...string) error
 	return nil
 }
 
+// DeleteRule is part of Interface.
```
Comment (smarterclayton) [inline pkg/util/iptables/iptables.go]: Mismatched godoc

**Comment by smarterclayton** [comment]
The common iptables changes look fine, as does the overall flow.  It might be good to get a reviewer who has used iptables more recently than me for the actual firewall creation parts.  

Is there a security risk for arbitrary clients in the cluster to contact the master such that we in the future would require some sort of shared secret?

**Comment by smarterclayton** [comment]
The client is vulnerable to hot looping if connection errors occur, would recommend having a jittered sleep interval in there (wait.Jitter)

**Comment by ArtfulCoder** [comment]
Thanks Clayton. It's easy to address the jitter feedback. 
Will think about the secure master comment.

**Comment by ArtfulCoder** [comment]
The clients contact the server which runs as a pod. Clients dont hit the master node.
to get firewall rules, any pod can as easily get it just like the server pod does.
So we are not increasing the security risk.

**Comment by ArtfulCoder** [comment]
@k8s-bot test this

**Comment by fejta** [comment]
This PR has had no meaningful activity for multiple months. If it is still valid please rebase, push a new commit and reopen the PR. Thanks!

**Pull request closed without merging**